policy/decompressor.go

```diff
@@ -0,0 +1,133 @@
+package policy
+
+import (
+	"sync"
+
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3metrics/generated/proto/schema"
+	"github.com/m3db/m3x/instrument"
+	"github.com/m3db/m3x/log"
+)
+
+// Decompressor maintains a mapping of integers to policies. It is used to return a Policy from
+// its corresponding integer encoding
+type Decompressor interface {
+	Get(id int64) (Policy, bool)
+}
+
+type decompressorMap map[int64]Policy
+
+type noopDecompression struct{}
+
+// NewNoopDecompressor returns a new Decompressor mapping which never matches an id
+func NewNoopDecompressor() Decompressor {
+	return noopDecompression{}
+}
+
+func (m noopDecompression) Get(id int64) (Policy, bool) {
+	return EmptyPolicy, false
+}
+
+type staticDecompressor struct {
+	policies decompressorMap
+}
+
+// NewStaticDecompressor returns a new static Decompressor
+func NewStaticDecompressor(policies map[int64]Policy) Decompressor {
+	return staticDecompressor{policies: policies}
+}
+
+func (m staticDecompressor) Get(id int64) (Policy, bool) {
+	p, ok := m.policies[id]
+	return p, ok
+}
+
+type dynamicDecompressor struct {
+	sync.RWMutex
+
+	policies decompressorMap
+	log      xlog.Logger
+}
+
+// NewDynamicDecompressor returns a new dynamic Decompressor
+func NewDynamicDecompressor(w kv.ValueWatch, opts instrument.Options) Decompressor {
+	m := &dynamicDecompressor{
+		policies: make(decompressorMap),
+		log:      opts.Logger(),
+	}
+	m.update(w.Get())
```

Comment:
Use `runtime/value.go`, which provides watching and unwatching functionality for you out of the box.